src/pyatsimagebuilder/builder.py

```diff
@@ -434,5 +436,28 @@ def _build_image(self, no_cache=False):
         if not self.image.id:
             # we've failed to set the image id - something is wrong!
             raise Exception('No confirmation of successful build.')
-    
 
+    def _replace_environment_variables(self):
+
+        ENV_PATTERN = re.compile(r'(%ENV{ *([0-9a-zA-Z\_]+) *})')
+
+        def _recursive_handle_leaf(data, handle):
+            if isinstance(data, dict):
+                for key in data:
+                    if isinstance(data[key], str):
+                        data[key] = handle(data[key])
+                    elif isinstance(data[key], list):
+                        for i in range(len(data[key])):
+                            data[key][i] = handle(data[key][i])
+                    else:
+                        _recursive_handle_leaf(data[key], handle)
+            else:
+                raise TypeError("Need dict, type={}".format(type(data)))
```

Comment:
In the case of an int or bool this will raise a TypeError. Handle dicts explicitly, not as the assumed default type. If a value isn't a string or a list or a dict, it can just be ignored.

```suggestion
        def _recursive_handle_leaf(data, handle):
            if isinstance(data, dict):
                for key, val in data.items():
                    if isinstance(val, str):
                        data[key] = handle(val)
                    elif isinstance(val, list):
                        for i in range(len(val)):
                            data[key][i] = handle(val[i])
                    elif isinstance(val, dict):
                        _recursive_handle_leaf(val, handle)
```

Reply:
done. ignores if not str, list, dict. also on list values, only handles if str